# RockingData Challenge

Marketing para un banco de Portugal

Usando anaconda env

> conda create --name marketing python=3.11

> conda install --yes --file requirements.txt


## Librerias

In [12]:
import numpy as np
import pandas as pd
import tqdm as notebook_tqdm
from ydata_profiling import ProfileReport

from utils import summary_statistics,\
                    summary_statistics_comparative,\
                    plot_distributions

## EDA

### Análisis express usando ydata_profiling

Esta librería procesa las variables del dataset, una a una, evaluando:

- Tipo de dato
- Estadísticos básicos (mean, min, max)
- Distribuciones
- Uniques, Missing, Infinite, Zeros
- Duplicados
- Interacciones o pairplots de numéricas
- Correlación entre numéricas de numéricas

Este tipo de herramientas permite un rápido vistazo, para luego profunzidar en lo que amerita.

Limitante: resulta lento o impracticable con un dataset de cientos de variables, lo que amerita evaluar la relevancia de variables de forma previa.

In [3]:
df_ori = pd.read_csv('data/bank-additional-full.csv', sep=';')

In [4]:
df_ori.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [5]:
profile = ProfileReport(df_ori, title="Bank Marketing")

In [6]:
profile.to_file("EDA_report_Bank_Marketing.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

## Diccionario de variables

**Bank client data:**

1 - age (numeric)

2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')

3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)

4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')

5 - default: has credit in default? (categorical: 'no','yes','unknown')

6 - housing: has housing loan? (categorical: 'no','yes','unknown')

7 - loan: has personal loan? (categorical: 'no','yes','unknown')

**Related with the last contact of the current campaign:**

8 - contact: contact communication type (categorical: 'cellular','telephone')

9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')

10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')

11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.

**Other attributes:**

12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)

13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)

14 - previous: number of contacts performed before this campaign and for this client (numeric)

15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')

**Social and economic context attributes**

16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)

17 - cons.price.idx: consumer price index - monthly indicator (numeric)

18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)

19 - euribor3m: euribor 3 month rate - daily indicator (numeric)

20 - nr.employed: number of employees - quarterly indicator (numeric)

**Output variable (desired target):**

21 - y - has the client subscribed a term deposit? (binary: 'yes','no')


## Data Quality

### Missing

En la información se aclara que hay missing values en variables categóricos. 

En realidad, el dataset provisto los tiene indicados como "unkown".

Estas son:

- marital < 1 %
- education 4 %
- default 21 %
- housing 2 %
- loan 2 %

In [7]:
### No se tomarán medidas con los missings, considerándolos como nuevas categorías

### Duplicates

Encontramos una mínima cantidad de missing values, que resulta práctico eliminar.

In [10]:
df_ori.drop_duplicates(inplace=True)

## Data Drifting

Al tratarse de un dataset atemporal, o una única foto extraída sin datos históricos, el data drifting no es un problema.

En futuras versiones del modelo, con info de los clientes en distintos períodos de tiempo (meses por ejemplo), esto debe ser revisado.

In [ ]:
### NO APLICA

## Feature engineering

La ingeniería de variables resulta un paso crucial en el desarrollo de modelos predictivos. En este challenge, devido a la escaza cantidad de features originales, avanzaré sin crear variables. 

> En un futuro, contar otros datos de los clientes, variables como: balance, ingresos, gastos, transacciones, deudas y productos bancarios; puede ser muy benificioso para mejorar el perfomance de la solución.

*Esto queda marcado como un punto de mejora a futuro*

In [ ]:
### MEJORA

## Target

La variable objetivo y, que parece indicar la conversión de la venta de "plazo fijo", tiene ~11 % de casos positivos y ~89 negativos.

Este desbalance no motiva, a priori, tomar ninguna medida de balanceo. Sin embargo, sí sesga métricas como el accuracy.

## Ventas vs. no-venta

Para entender las condiciones de venta vs. los casos donde las campaña no fue exitosa, se procede a visualizar los datos según la variable target.

In [16]:
target = 'y' 

print("Resumen Estadístico - Comparativo")
key1= 'yes'
key2= 'no'
stats=['count', 'mean', '50%', 'std']
summary_stats_combined = summary_statistics_comparative(df_ori.loc[df_ori[target] == key1], 
                                                        df_ori.loc[df_ori[target] == key2], 
                                                        key1, key2, stats=stats)
print(summary_stats_combined.to_markdown())


Resumen Estadístico - Comparativo
|                |   ('yes', 'count') |   ('no', 'count') |   ('yes', 'mean') |   ('no', 'mean') |   ('yes', '50%') |   ('no', '50%') |   ('yes', 'std') |   ('no', 'std') |
|:---------------|-------------------:|------------------:|------------------:|-----------------:|-----------------:|----------------:|-----------------:|----------------:|
| age            |               4639 |             36537 |         40.9123   |        39.911    |           37     |          38     |        13.8388   |        9.89718  |
| duration       |               4639 |             36537 |        553.256    |       220.868    |          449     |         164     |       401.191    |      207.116    |
| campaign       |               4639 |             36537 |          2.05195  |         2.63339  |            2     |           2     |         1.66635  |        2.87377  |
| pdays          |               4639 |             36537 |        791.991    |       984.109    |   

In [ ]:
### pr

## Segmentación de las ventas

Para un mejor entendimiento del caso, se procede a segmentar los casos exitosos, para entender las características de los clientes y el contexto en el cuál